### PR TITLE
Add initial payload generators

### PR DESCRIPTION
**Pull request by sbaier1**
MV-33

### CODE REVIEW

File: src/main/java/org/varroa/chunk/command/generation/PublishPayloadGeneration.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        return payloadGeneratorType;
+    }
+
+    public @NotNull String getMessage() {
+        return message;
+    }
+
+    public @Nullable PayloadGenerator createGenerator(final int clientOffset, final @NotNull String clientId) {
+        // Some simple generators can be offset agnostic. With meaningful payloads, most aren't though.
+        if (payloadGeneratorType == GENERATOR_TYPE_RANDOM) {
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/chunk/command/generation/PublishPayloadGeneration.java]: a switch would be nicer here

File: src/main/java/org/varroa/chunk/command/generation/PublishPayloadGeneration.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        }
+
+        public @NotNull Builder message(final @NotNull String message) {
+            this.message = message;
+            return this;
+        }
+
+        public @NotNull PublishPayloadGeneration build() {
+            if(message == null) {
+                log.error("No payload message specified in payload generation builder");
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/chunk/command/generation/PublishPayloadGeneration.java]: please throw an exception

File: src/main/java/org/varroa/chunk/command/generation/PublishPayloadGeneration.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            }
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/chunk/command/generation/PublishPayloadGeneration.java]: `type != -1` could be checked here

File: src/main/java/org/varroa/load/Mqtt5Agent.java [hunk 4]
```diff
@@ -85,19 +92,24 @@
             // we publish only one message concurrently
             publishResultFlowable =
                     rateFlowable.flatMap(
-                            offset -> client.publish(Flowable.just(getPublish(topicNameGeneration, qos))), 1);
+                            offset -> client.publish(Flowable.just(getPublish(topicNameGeneration, qos, generator, publishCommand))), 1);
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/load/Mqtt5Agent.java]: Please enable automatic formatting on commit, the lines are to long for reading

File: src/main/java/org/varroa/load/Mqtt5Agent.java [hunk 5]
```diff
@@ -85,19 +93,27 @@
...
 
         return publishResultFlowable.ignoreElements();
     }
 
     private @NotNull Mqtt5Publish getPublish(
-            final @NotNull TopicNameGeneration topicNameGeneration, final @NotNull MqttQos qos) {
-
-        return Mqtt5Publish.builder().topic(getTopic(topicNameGeneration)).qos(qos).build();
+            final @NotNull TopicNameGeneration topicNameGeneration, final @NotNull MqttQos qos,
+            final @NotNull PayloadGenerator generator,
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/load/Mqtt5Agent.java]: minor: `generator` -> `payloadGenerator`

File: src/main/java/org/varroa/remote/codec/encoder/EncodeUtils.java [hunk 6]
```diff
@@ -128,6 +129,19 @@ public static void encodeTopicGeneration(
...
     }
 
+    public static void encodePublishPayloadGeneration(
+            final @NotNull ByteBuf out,
+            final @NotNull PublishPayloadGeneration generation) {
+        encodeString(out, generation.getMessage());
+        out.writeInt(generation.getPayloadGeneratorType());
+    }
+
+    public static int encodePublishPayloadGenerationLength(
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/remote/codec/encoder/EncodeUtils.java]: minor: encodePublishPayloadGenerationLength -> encode**d**PublishPayloadGenerationLength

File: src/main/java/org/varroa/scenario/payload/PayloadGenerator.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
+package org.varroa.scenario.payload;
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/scenario/payload/PayloadGenerator.java]: Is this the right package, all other generators are located in `org.varroa.chunk.command.generation`
maybe move the PayloadGenerators to `org.varroa.chunk.command.generation.payload`?

File: src/main/java/org/varroa/scenario/xml/adapters/PayloadGeneratorTypeAdapter.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+public class PayloadGeneratorTypeAdapter extends XmlAdapter<String, Integer> {
+    private static final String ALPHA_NUMERIC_GENERATOR = "randomAlphaNumeric";
+    private static final String REGEX_GENERATOR = "regex";
+    private static final String MUSTACHE_GENERATOR = "mustache";
+    private static final String STATIC_GENERATOR = "static";
+    private static final String UNSPECIFIED_GENERATOR = "unspecified";
+
+    @Override
+    public @NotNull Integer unmarshal(@NotNull String v) throws Exception {
+        if (ALPHA_NUMERIC_GENERATOR.equals(v)) {
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/scenario/xml/adapters/PayloadGeneratorTypeAdapter.java]: switch case would be nicer

File: src/main/java/org/varroa/scenario/xml/adapters/PayloadGeneratorTypeAdapter.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        } else if (REGEX_GENERATOR.equals(v)) {
+            return PublishPayloadGeneration.GENERATOR_TYPE_REGEX;
+        } else {
+            return PublishPayloadGeneration.GENERATOR_TYPE_STATIC;
+        }
+    }
+
+    @Override
+    public @NotNull String marshal(@NotNull Integer v) {
+        if (v == PublishPayloadGeneration.GENERATOR_TYPE_RANDOM) {
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/scenario/xml/adapters/PayloadGeneratorTypeAdapter.java]: switch would be nicer

File: src/main/resources/scenario.xsd [hunk 10]
```diff
@@ -207,6 +207,7 @@
             <xs:extension base="command">
                 <xs:attribute name="topicGroup" type="xs:IDREF" use="required"/>
                 <xs:attribute name="message" type="xs:string" use="required"/>
+                <xs:attribute name="payloadGeneratorType" type="xs:string"/>
```
Comment (SgtSilvio) [inline src/main/resources/scenario.xsd]: we should provide an enum in the xsd for the available types, maybe also add a default value of "static"

**Comment by SgtSilvio** [comment]
very nice PR 👍

File: src/main/java/org/varroa/scenario/payload/PayloadGenerator.java (hunk 7 above)
Comment (sbaier1) [reply src/main/java/org/varroa/scenario/payload/PayloadGenerator.java]: makes sense, will do

File: build.gradle [hunk 11]
```diff
@@ -51,6 +51,7 @@ dependencies {
             [group: 'javax.xml.bind', name: 'jaxb-api', version: '2.3.1'],
             [group: 'org.glassfish.jaxb', name: 'jaxb-runtime', version: '2.3.2'],
             [group: 'com.beust', name: 'jcommander', version: '1.72'],
+            [group: 'com.github.spullara.mustache.java', name: 'compiler', version: '0.9.6'],
```
Comment (sauroter) [inline build.gradle]: where do you find these dependencies 😀 spullara, mifmif ...

File: src/main/java/org/varroa/chunk/command/generation/payload/MustachePayloadGenerator.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    public @NotNull byte[] nextPayload(
+            final @NotNull PublishCommand publishCommand,
+            final @NotNull String topic) {
+        // Initial simple scope. This may be customized in the future
+        final Map<String, Object> scope = new HashMap<>();
+        scope.put("topic", topic);
+        scope.put("clientId", clientId);
+        final StringWriter writer = new StringWriter();
+        mustache.execute(writer, scope);
+        return writer.getBuffer().toString().getBytes();
```
Comment (sauroter) [inline src/main/java/org/varroa/chunk/command/generation/payload/MustachePayloadGenerator.java]: is it possible to write directly to an byte array or bytebuf? I dont like that we are pulling the generated payload into memory.

File: src/main/java/org/varroa/chunk/command/generation/payload/MustachePayloadGenerator.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+ */
+public class MustachePayloadGenerator implements PayloadGenerator {
+
+    private final @NotNull Mustache mustache;
+    private final @NotNull String clientId;
+
+    public MustachePayloadGenerator(final @NotNull String clientId, final @NotNull String message) {
+        final DefaultMustacheFactory mustacheFactory = new DefaultMustacheFactory();
+        this.clientId = clientId;
+        this.mustache = mustacheFactory.compile(new StringReader(message), "pattern");
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/chunk/command/generation/payload/MustachePayloadGenerator.java]: The compiled pattern should probably be cached

File: src/main/java/org/varroa/chunk/command/generation/payload/MustachePayloadGenerator.java (hunk 12 above)
Comment (SgtSilvio) [reply src/main/java/org/varroa/chunk/command/generation/payload/MustachePayloadGenerator.java]: good point
maybe use a OutputStreamWriter which writes to a io.netty.buffer.ByteBufOutputStream
the hivemq-mqtt-client uses ByteBuffer for payloads and can benefit from direct ByteBuffer

Comment (sbaier1) [reply src/main/java/org/varroa/chunk/command/generation/payload/MustachePayloadGenerator.java]: yeah this was something i also didn't like very much, i'll look into it, liking @SgtSilvio 's idea

**Pull request merged**